types/reconnect-core/index.d.ts

```diff
@@ -0,0 +1,47 @@
+// Type definitions for reconnect-core 1.3
+// Project: https://www.npmjs.com/package/reconnect-core
+// Definitions by: James Bromwell <https://github.com/thw0rted>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// TypeScript Version: 2.8
+
+import { Backoff, ExponentialOptions  } from "backoff";
+import { EventEmitter } from "events";
+import { Stream } from "stream";
+
+export namespace reconnect {
+    // TODO: Once DT supports TS 3.0, use new Generic Rest Parameter feature for
+    // ArgType.  Current definitions only support one parameter passed to
+    // connect() but the library actually allows any number of args.
+    type CustomModule<ArgType, ConnectionType> =
+        (opts?: ModuleOptions<ConnectionType>, cb?: (con: ConnectionType) => void) => Instance<ArgType, ConnectionType>;
+
+    interface ModuleOptions<ConnectionType> extends ExponentialOptions {
+        immediate?: boolean;
+        failAfter?: number;
+        onConnect?: (con: ConnectionType) => void;
+        strategy?: "fibonacci" | "exponential" | Backoff;
+    }
+
+    interface Instance<ArgType, ConnectionType> extends EventEmitter {
+        connect(opts: ArgType): this;
+        listen(opts: ArgType): this;
+        disconnect(): this;
+        reset(): void;
+
+        readonly connected: boolean;
+        reconnect: boolean;
+
+        on(event: "backoff", cb: (n: number, delay: number, err?: any) => void): this;
+        on(event: "connect" | "connection", cb: (con: ConnectionType) => void): this;
+        on(event: "disconnect", cb: (err?: any) => void): this;
+        on(event: "error" | "fail", cb: (err: any) => void): this;
+        on(event: "reconnect", cb: (n: number, delay: number) => void): this;
+    }
+}
+
+type ConnectFunction<ArgType, ConnectionType> =
+    (this: reconnect.Instance<ArgType, ConnectionType>, opts: ArgType) => ConnectionType;
+export function reconnect<ArgType, ConnectionType>(
+    cf: ConnectFunction<ArgType, ConnectionType>): reconnect.CustomModule<ArgType, ConnectionType>;
+
+export default reconnect;
```

Comment:
It seems like the underlying package does not use `export default`.
Given that, you shouldn't use `export default` here either.
Instead, you should follow [this template](http://www.typescriptlang.org/docs/handbook/declaration-files/templates/module-function-d-ts.html).

Reply:
Honestly, I was never clear on exactly how this worked, but maybe you can help me.  If I `export` the function and namespace, then also `export default`, I can `import reconnect from "reconnect-core"` and the generated code works.  If I `declare namespace`, `declare function`, then `export = reconnect` (per your linked template), the only way to use it is either `import * as reconnect` or `import reconnect = require(...)`.

I typically expect to `import foo from ...` when using single-purpose modules like this and it would be nice to retain the behavior.  Is it possible without `export default`?  Also, since the JS doesn't specify a default, why does `export default` in the declaration file still generate working code?

Reply:
Sure thing, it's super confusing (and I think I'm still figuring it out, so forgive me if I get something wrong myself). As discussed [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import#Importing_defaults), `import name from 'module';` specifically works when the library itself uses `export default`. Using [unpkg to inspect reconnect-core](https://unpkg.com/reconnect-core@1.3.0/index.js), it looks like it is just a callable module, with no default export, so in this case, `import reconnect from "reconnect-core"` is technically wrong.

Next question: why does it work when you do the thing I'm recommending against? It has to do with the settings under which you're consuming the code. In TS 2.7, we introduced the [`esModuleInterop`](https://github.com/Microsoft/TypeScript/wiki/What%27s-new-in-TypeScript#support-for-import-d-from-cjs-form-commonjs-modules-with---esmoduleinterop) flag, which makes this work. But not everyone consuming your types will necessarily have this flag on, and it is better to have the type declaration file actually reflect the underlying library than to make assumptions about what other .d.ts consumers may or may not have enabled. Note that with this flag, you will be able to use `import reconnect from "reconnect-core";`, despite there being no `export default` in either the underlying library or in the type declaration file.

Last question: When changing the .d.ts file so that it uses `declare namespace reconnect` and `export = reconnect`, then I can access the types and interfaces in the namespace:
![image](https://user-images.githubusercontent.com/9092011/44162721-f792a200-a075-11e8-81e9-f7e521a85ac9.png)
If you want `ConnectFunction` to be accessible as well, you should put it in the namespace.

Please let me know if you have any more questions!

Reply:
Also, it seems like when you fix up the .d.ts file as I suggested, you won't need to disable those two lint rules anymore.

Reply:
...I *totally* forgot that I have `esModuleInterop` turned on.  Maybe it should be forcibly disabled for all of DT, to encourage definitions that match the actual package?  Anyway, I've made the changes you suggested. Thanks for all your help!